File: ruby/dry.rb
# DRY (Don't Repeat Yourself)
# If a piece of code is identical or very similar to another, try to extract it into a generalized function

# Before
def greet_morning
  puts "Good morning, Alice!"
end

def greet_afternoon
  puts "Good afternoon, Alice!"
end

def greet_evening
  puts "Good evening, Alice!"
end

greet_morning
greet_afternoon
greet_evening


# After
def greet(day_period, name)
  puts "Good #{day_period}, #{name}!"
end

greet("morning", "Alice")
greet("afternoon", "Alice")
greet("evening", "Alice")

File: csharp/Dry.cs
// DRY (Don't Repeat Yourself)
// If a piece of code is identical or very similar to another, try to extract it into a generalized function

using System;

public class Dry
{
    // Before
    public static void GreetMorning()
    {
        Console.WriteLine("Good morning, Alice!");
    }

    public static void GreetAfternoon()
    {
        Console.WriteLine("Good afternoon, Alice!");
    }

    public static void GreetEvening()
    {
        Console.WriteLine("Good evening, Alice!");
    }

    // After
    public static void Greet(string dayPeriod, string name)
    {
        Console.WriteLine($"Good {dayPeriod}, {name}!");
    }

    public static void Main(string[] args)
    {
        // Before usage
        GreetMorning();
        GreetAfternoon();
        GreetEvening();

        Console.WriteLine();

        // After usage
        Greet("morning", "Alice");
        Greet("afternoon", "Alice");
        Greet("evening", "Alice");
    }
}
